src/contexts/addresses.tsx

```diff
@@ -0,0 +1,172 @@
+
+  const updateAddressState = useCallback((newAddressState: AddressState) => {
+    setAddressesState((prevState) => {
+      const newAddressesState = new Map(prevState)
+      newAddressesState.set(newAddressState.hash, newAddressState)
+      console.log('updating state of address: ', newAddressState.hash)
+      return newAddressesState
+    })
+  }, [])
+
+  const fetchAddressDetails = useCallback(
+    async (addressHash: string) => {
+      if (!client) return
+      const addressDetails = await client.explorer.getAddressDetails(addressHash)
+      return addressDetails.data
+    },
+    [client]
+  )
+
+  const refreshAddressesDetails = async () => {
+    const newAddressesState = new Map(addressesState)
+    for (const [hash, addressState] of addressesState) {
+      const addressDetails = await fetchAddressDetails(hash)
+      if (addressDetails) {
+        newAddressesState.set(hash, { ...addressState, details: addressDetails })
+      }
+    }
+    setAddressesState(newAddressesState)
+  }
+
+  const saveNewAddress = useCallback(
+    async (newAddress: AddressState) => {
+      storeAddressIndexAndSettingsOfAccount(currentUsername, newAddress.index, newAddress.settings)
+
+      if (!newAddress.details) {
+        const addressDetails = await fetchAddressDetails(newAddress.hash)
+        updateAddressState({ ...newAddress, details: addressDetails })
+      } else {
+        updateAddressState(newAddress)
+      }
+    },
+    [currentUsername, fetchAddressDetails, updateAddressState]
+  )
+
+  useEffect(() => {
+    const initializeAddresses = async () => {
+      if (!currentUsername || !wallet) return
+
+      const addressesIndexAndSettings = loadStoredAddressesIndexAndSettingsOfAccount(currentUsername)
+
+      if (addressesIndexAndSettings.length === 0) {
+        await saveNewAddress({
+          hash: wallet.address,
+          publicKey: wallet.publicKey,
+          privateKey: wallet.privateKey,
+          group: addressToGroup(wallet.address, TOTAL_NUMBER_OF_GROUPS),
+          index: 0,
+          settings: {
+            isMain: true,
+            label: undefined,
+            color: undefined
+          }
+        })
+      } else {
+        for (const { index, ...settings } of addressesIndexAndSettings) {
+          const { address, publicKey, privateKey } = deriveNewAddressData(wallet.seed, undefined, index)
+          await saveNewAddress({
+            hash: address,
+            publicKey,
+            privateKey,
+            group: addressToGroup(address, TOTAL_NUMBER_OF_GROUPS),
+            index,
+            settings: settings
+          })
+        }
+      }
+    }
+
+    initializeAddresses()
+  }, [currentUsername, saveNewAddress, wallet])
+
+  // Clean state when locking the wallet
+  useEffect(() => {
+    if (wallet === undefined) {
+      setAddressesState(new Map())
+    }
+  }, [wallet])
+
+  return (
+    <AddressesContext.Provider
+      value={merge(
```

Comment:
Hm, it's not an array, but a Map. Is Map also not considered a deep object (its key is a string and its value is deep objects)? I'll check it.

Reply:
You're right, sorry. But my point is still partially valid: if only the map needs a deep merge, we could restrict the operation to this field only.

But honestly, this might be over engineering at this point, so maybe just something to keep in mind in the future :)